### PR TITLE
Add default toolchains for MacOS.

### LINKED ISSUE #1096: Add a timeout as a generic feature of a task.
During test run executions, there is a possibility of tests hanging indefinitely.

### Expected Behavior

User must be able to set time outs for tasks(at least for tasks of type: Test)
### Current Behavior

No provision to set timeouts.
### Context

We queue tests to run in our lab vm's and we have seen situations wherein the lab machines couldn't pick up new tests because of some issues in the tests that got picked up previously and since there is no time outs that can be applied similar to ant's limit task.

### Your Environment

 * Build scan URL:

### PR DESCRIPTION
**Pull request by prbprbprb**
Fixes [gradle-native #1096](https://github.com/gradle/gradle-native/issues/1096)

Adds two new default toolchain targets for MacOS x86_64 and
arm64 which use Clang's `-arch` argument to target the correct
architecture.  On MacOS, this is used in preference to the
existing `Intel64Architecture` target which passes `-m64` to the
toolchain.

This is sufficient for `NativeExecutableSpec` to target
`osx_x86-64` and `osx_aarch64` on both Intel and M1 Macs.

I haven't (yet!) located any tests for this functionality, and
so this PR doesn't include any test changes.  Happy to add them
if you can point me in the right direction.

Signed-off-by: Pete Bentley <prb@google.com>

Fixes [gradle-native #1096](https://github.com/gradle/gradle-native/issues/1096)

### Context
Ease of use on both Intel and ARM based Macs.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by prbprbprb** [comment]
Odd, DCO is showing as failing but the single commit in this PR has a correct `Signed-off-by` tag.

**Comment by prbprbprb** [comment]
@big-guy possibly of interest to you as you recently fixed the arm-v8/aarch64 confusion in native builds?

**Comment by big-guy** [comment]
@bot-gradle test this

**Comment by big-guy** [comment]
Thanks @prbprbprb 

We have an experimental M1-based pipeline I can run this through

**Comment by prbprbprb** [comment]
Thanks! I kinda forgot about the docs changes, I'll get on that.

And do you have any pointer to where I should be adding tests?

**Comment by big-guy** [comment]
@prbprbprb take a look at https://github.com/gradle/gradle/blob/89ea0aa47a16cc0b2e741154476d0847a9c4161a/subprojects/platform-native/src/test/groovy/org/gradle/nativeplatform/toolchain/internal/gcc/AbstractGccCompatibleToolChainTest.groovy#L312

I think making that also check for aarch64 would do it.

I think these are the only docs that need to be updated right now:
https://github.com/gradle/gradle/blob/34f6ad81c5f95c3da29eb7a6e273b00344d75b17/subprojects/docs/src/docs/userguide/native/native_software.adoc#using-tool-chains

I think there are more things we can do after this, but this is enough to get this merged.

Thanks!

**Comment by prbprbprb** [comment]
Thanks, I'll work through that!  One corner case I forgot - if there are versions of Xcode in the wild old enough not to support the `-arch` probe args we might need to fall back to `-m64`.  I'll talk to some MacOS folk and see :)

Happy to take follow-on work too as we also want to support ARM Linux better in Conscrypt.

**Comment by prbprbprb** [comment]
I do apologise, kind of swamped with the day job recently. I will have time to look at this PR over the coming weekend but if it's easier for you take over the work in the meantime then you have my blessing.

**Comment by prbprbprb** [comment]
Working on it but note that both the `supplies args` tests in `AbstractGccCompatibleToolChainTest.groovy` appear to be currently broken... I can randomly change the expectations in the `where:` clause and they still pass :)

**Comment by prbprbprb** [comment]
Unit tests and docs updated.  Do we need an integration test?

**Comment by prbprbprb** [comment]
> do you think you'll have time to [...] look at the corner-case for older Xcode versions?

Looks like the `-arch` flag has been around on MacOS since at least their [PPC days](https://developer.apple.com/library/archive/technotes/tn2005/tn2137.html), so I don't think we need to worry about it.

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/gcc/AbstractGccCompatibleToolChain.java
Comment (lacasseio) [inline subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/gcc/AbstractGccCompatibleToolChain.java:97]: 💄 It would be better if we could avoid duplicated architecture. It would be possible to choose different flags by using [getPlatform()](https://github.com/gradle/gradle/blob/5ec3f672ed600a86280be490395d70b7bc634862/subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/gcc/DefaultGccPlatformToolChain.java#L80) on the toolchain. The downside of merging `Intel64Architecture` and `OsxIntelArchitecture` is some tests may need to be adjusted which can be troublesome.

Comment (DPUkyle) [reply subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/gcc/AbstractGccCompatibleToolChain.java:97]: I think you're suggesting three architectures, correct? And the probing for OS-dependent qualities will happen in their implementations?

- Intel32Architecture
- Intel64Architecture
- Arm64Architecture

Comment (lacasseio) [reply subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/gcc/AbstractGccCompatibleToolChain.java:97]: We could use `Arm64Architecture`, but I think for this PR using `OsxArm64Architecture` would be enough. Adding `arm64` support for all OS is a big can of worm to open.

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@lacasseio can you have another look?

**Comment by big-guy** [comment]
@bot-gradle help

**Review by nathan-contino — APPROVED** [review]
LGTM

**Comment by DPUkyle** [comment]
@bot-gradle test and merge

**Pull request merged**
